### ISSUE TITLE
训练时出现DEROUTPUT dimensions are incompatible with X and FILTERS

### ISSUE THREAD

**Issue by weidelongdongqiang** [zwx8981/BIQA_project#6]
大神您好！同前几个同学一样，我在运行run_exp.m的时候也出现这个问题。
错误使用 vl_nnconv
DEROUTPUT dimensions are incompatible with X and FILTERS.

出错 vl_simplenn (line 413)
          vl_nnconv(res(i).x, l.weights{1}, l.weights{2}, res(i+1).dzdx, ...

出错 cnn_train>processEpoch (line 350)
    res = vl_simplenn(net, im, dzdy, res, ...

出错 cnn_train (line 150)
    [net, state] = processEpoch(net, state, params, 'train') ;

出错 initializeNetworkTwoStreams (line 280)
        [netc, info] = cnn_train(netc, bcnndb, @getBatch_bcnn_fromdisk, opts.inittrain, ...

出错 imdb_bcnn_train_dag (line 70)
net = initNetFn(imdb, encoderOpts, opts);

出错 run_experiments_bcnn_train (line 36)
          imdb_bcnn_train_dag(imdb, options);

出错 run_exp (line 63)
    [options, imdb] = run_experiments_bcnn_train(opts);
我看您没有回答。我把读取图像的路径改成我自己的了，matconvnet与你版本是一样的，另外我是单GPU，把那个编号由2改为1了，其它没有动。我GPU是1070ti，请大神指教！谢谢！

**Comment by zwx8981** [comment]
@weidelongdongqiang 
你好，这应该是因为vl_nnpdist要求预测值跟标签具有相同的形状，这个地方我在我自己的matconvnet中的vl_simplenn里做过一个修改：

    case 'pdist'
      l.class = single(reshape(l.class,1,1,1,[]));
      res(i+1).x = vl_nnpdist(res(i).x, l.class, l.p, ...
        'noRoot', l.noRoot, ...
        'epsilon', l.epsilon, ...
        'aggregate', l.aggregate, ...
        'instanceWeights', l.instanceWeights) ; 

同样的问题，在DagNN的封装PDist.m文件中也需要做相应的修改:
%PDIST vl_nnpdist dagnn wrapper
%  Accepts 2 or 3 inputs, where third input is used as variable
%  'instanceWeights' parameter. Derivatives for the 3rd input are not
%  computed.
%  By default aggregates the element-wise loss.
classdef PDist < dagnn.Loss
  properties
    p = 2;
    aggregate = true;
    noRoot = false;
  end

  methods
    function outputs = forward(obj, inputs, params)
      switch numel(inputs)
        case 2
            inputs{2} = reshape(inputs{2},1,1,1,[]);
          outputs{1} = vl_nnpdist(inputs{1}, inputs{2}, obj.p, [], ...
            'aggregate', obj.aggregate, obj.opts{:},'noRoot',obj.noRoot) ;
        case 3
          outputs{1} = vl_nnpdist(inputs{1}, inputs{2}, obj.p, [], ...
            'aggregate', obj.aggregate, 'instanceWeights', inputs{3}, ...
            obj.opts{:}) ;
        otherwise
          error('Invalid number of inputs');
      end
      obj.accumulateAverage(inputs, outputs);
    end

    function [derInputs, derParams] = backward(obj, inputs, params, derOutputs)
      derInputs = cell(1, numel(inputs));
      switch numel(inputs)
        case 2
            inputs{2} = reshape(inputs{2},1,1,1,[]);
          [derInputs{1}, derInputs{2}] = vl_nnpdist(inputs{1}, inputs{2}, ...
            obj.p, derOutputs{1}, 'aggregate', obj.aggregate, obj.opts{:},'noRoot',obj.noRoot) ;
        case 3
          [derInputs{1}, derInputs{2}] = vl_nnpdist(inputs{1}, inputs{2}, ...
            obj.p, derOutputs{1}, 'aggregate', obj.aggregate, ...
            'instanceWeights', inputs{3}, obj.opts{:}) ;
        otherwise
          error('Invalid number of inputs');
      end
      derParams = {} ;
    end

    function obj = PDist(varargin)
      obj.load(varargin) ;
      obj.loss = 'pdist';
    end
  end
end

**Comment by zwx8981** [comment]
@weidelongdongqiang 我上传了修改后的文件到matconvet/matlab

**Comment by weidelongdongqiang** [comment]
非常感谢！问题解决了。
